### ISSUE TITLE
Variable shadowing is reported unsupported when associated type is used

### ISSUE THREAD

**Issue by euclidgame** [verus-lang/verus#705]
I wrote a `proof` function which is roughly like:
```
pub proof fn lemma_x<K: ResourceView, R: Reconciler<K>>(
    spec: TempPred<State<K, R>>, cr: K, state: R::T, next_state: R::T
)
    requires
        ..., // some preconditions
    ensures
        ..., // some postconditions
{
    let filter = |s: R::T| { s == next_state };
    temp_pred_equality(
        lift_state(at_reconcile_state::<K, R>(cr.object_ref(), next_state)),
        lift_state(at_expected_reconcile_states::<K, R>(cr.object_ref(), filter))
    );
    lemma_y::<K, R>(spec, cr, state, filter);
}
```
Here the trait `Reconciler` takes a generic `K` and the trait also has an associated type `T` which is the type of the two parameters `state`, `next_state`.

And the verifier complains the following:
```
error: variable shadowing not yet supported
   --> kubernetes_cluster/proof/controller_runtime_liveness.rs:323:28
    |
323 |     let filter = |s: R::T| { s == next_state };
    |                            ^^^^^^^^^^^^^^^^^^^
```
I am a bit confused because I can't see variable shadowing here.

Plus, we previously made `T` also a generic to `Reconciler` and at that time the verification went through. The verification fails after we change it into an associated type.

We need some help on this. Please let me know if you need more information.

**Comment by tjhance** [comment]
Here's a self-contained example that exhibits the same problem:

```rust
pub trait Tr {
    type T;
}

pub proof fn foo<R: Tr>(next_state: R::T) {
    let filter = |s: R::T| { s == next_state };
}
```

@Chris-Hawblitzel can you take a look? The error comes from triggers.rs, but the problem also has to do with associated types, and I dunno what the interaction could be there

**Comment by Chris-Hawblitzel** [comment]
This bug was fixed by 913b24151df319358b819941e1e477b7e2d86a26 .  Before that, there was some dead code in ast_to_sst.rs's handling of ExprX::Closure that over time had become broken.  Adding associated types accidentally reactivated the broken dead code, causing this bug, until 913b24151df319358b819941e1e477b7e2d86a26 deactivated the broken dead code again.  Commit 421a5f6322101956b5adebc8b8b9b9b656201afb removes the broken dead code for good.
